core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java

```diff
@@ -79,135 +80,221 @@ public Result prepare(UpdateRequest request, IndexShard indexShard, LongSupplier
     }
 
     /**
-     * Prepares an update request by converting it into an index or delete request or an update response (no action).
+     * Prepares an update request by converting it into an index or delete request or an update response (no action, in the event of a
+     * noop).
      */
     @SuppressWarnings("unchecked")
     protected Result prepare(ShardId shardId, UpdateRequest request, final GetResult getResult, LongSupplier nowInMillis) {
-        if (!getResult.isExists()) {
+        if (getResult.isExists() == false) {
+            // If the document didn't exist, execute the update request as an upsert
+            return prepareUpsert(shardId, request, getResult, nowInMillis);
+        } else if (getResult.internalSourceRef() == null) {
+            // no source, we can't do anything, throw a failure...
+            throw new DocumentSourceMissingException(shardId, request.type(), request.id());
+        } else if (request.script() == null && request.doc() != null) {
+            // The request has no script, it is a new doc that should be merged with the old document
+            return prepareUpdateIndexRequest(shardId, request, getResult, request.detectNoop());
+        } else {
+            // The request has a script (or empty script), execute the script and prepare a new index request
+            return prepareUpdateScriptRequest(shardId, request, getResult, nowInMillis);
+        }
+    }
+
+    /**
+     * Execute a scripted upsert, where there is an existing upsert document and a script to be executed. The script is executed and a new
+     * Tuple of operation and updated {@code _source} is returned.
+     */
+    Tuple<UpdateOpType, Map<String, Object>> executeScriptedUpsert(IndexRequest upsert, Script script, LongSupplier nowInMillis) {
+        Map<String, Object> upsertDoc = upsert.sourceAsMap();
+        Map<String, Object> ctx = new HashMap<>(3);
+        // Tell the script that this is a create and not an update
+        ctx.put(ContextFields.OP, UpdateOpType.CREATE.toString());
+        ctx.put(ContextFields.SOURCE, upsertDoc);
+        ctx.put(ContextFields.NOW, nowInMillis.getAsLong());
+        ctx = executeScript(script, ctx);
+
+        UpdateOpType operation = UpdateOpType.lenientFromString((String) ctx.get(ContextFields.OP), logger, script.getIdOrCode());
+        Map newSource = (Map) ctx.get(ContextFields.SOURCE);
+
+        if (operation != UpdateOpType.CREATE && operation != UpdateOpType.NONE) {
+            // Only valid options for an upsert script are "create" (the default) or "none", meaning abort upsert
+            logger.warn("Invalid upsert operation [{}] for script [{}], doing nothing...", operation, script.getIdOrCode());
```

Comment:
This also seems like the kind of leniency that we'd want to remove in the future.